server/app/services/applicant/ServiceAreaUpdateResolver.java

```diff
@@ -0,0 +1,187 @@
+                newServiceAreaInclusionGroup =
+                    Stream.of(existingServiceAreaInclusionGroup, serviceAreaInclusionGroup)
+                        .flatMap(List::stream)
+                        .collect(
+                            Collectors.toMap(
+                                ServiceAreaInclusion::getServiceAreaId,
+                                area -> area,
+                                (ServiceAreaInclusion existingInclusion,
+                                    ServiceAreaInclusion newInclusion) ->
+                                    newInclusion == null ? existingInclusion : newInclusion))
+                        .values()
+                        .stream()
+                        .collect(ImmutableList.toImmutableList());
+              } else {
+                newServiceAreaInclusionGroup = serviceAreaInclusionGroup;
+              }
+
+              ServiceAreaUpdate serviceAreaUpdate =
+                  ServiceAreaUpdate.create(serviceAreaPath, newServiceAreaInclusionGroup);
+
+              return CompletableFuture.completedFuture(Optional.of(serviceAreaUpdate));
+            },
+            httpExecutionContext.current());
+  }
+
+  private Boolean doesUpdateContainCorrectedAddress(
+      ApplicantQuestion addressQuestion, ImmutableMap<String, String> updateMap) {
+    Path correctedPath = addressQuestion.getContextualizedPath().join(Scalar.CORRECTED);
+    String correctedValue = updateMap.get(correctedPath.toString());
+
+    if (correctedValue == null) {
+      return false;
+    }
+
+    return correctedValue.equals(CorrectedAddressState.CORRECTED.getSerializationFormat());
+  }
+
+  private ImmutableList<ServiceAreaInclusion> getExistingServiceAreaInclusionGroup(
+      Path serviceAreaPath, ImmutableMap<String, String> updateMap) {
+    String serviceAreaValue = updateMap.get(serviceAreaPath.toString());
+    ImmutableList.Builder<ServiceAreaInclusion> existingServiceAreaInclusionGroupBuilder =
+        ImmutableList.builder();
+    if (serviceAreaValue != null) {
+      existingServiceAreaInclusionGroupBuilder.addAll(
+          ServiceAreaInclusionGroup.deserialize(serviceAreaValue));
+    }
+
+    return existingServiceAreaInclusionGroupBuilder.build();
+  }
+
+  private AddressLocation getAddressLocationFromUpdates(
+      ApplicantQuestion addressQuestion, ImmutableMap<String, String> updateMap) {
+    Path latitudePath = addressQuestion.getContextualizedPath().join(Scalar.LATITUDE);
+    Path longitudePath = addressQuestion.getContextualizedPath().join(Scalar.LONGITUDE);
+    Path wellKnownIdPath = addressQuestion.getContextualizedPath().join(Scalar.WELL_KNOWN_ID);
+
+    return AddressLocation.builder()
+        .setLatitude(Double.parseDouble(updateMap.get(latitudePath.toString())))
+        .setLongitude(Double.parseDouble(updateMap.get(longitudePath.toString())))
+        .setWellKnownId(Integer.parseInt(updateMap.get(wellKnownIdPath.toString())))
+        .build();
```

Comment:
**question (blocking):** the address question form doesn't have these values in it and I think Gwen was planning on storing them in the session rather than in the DOM. Have y'all synced on that?

If we're going to go with storing them in the DOM we need to include and validate an HMAC.

Reply:
Good call. I'll make sure we're inlined on this before making any changes.

Reply:
@bion, looking at @gwendolyngoetz's latest changes to https://github.com/civiform/civiform/pull/4236, an update map is built from the session and passed into `stageAndUpdateIfValid`, so I think we can safely merge this.